### DISCUSSION TITLE
Specific questions about noise, JPG vs. PNG, and light-relative normals

### DISCUSSION THREAD

**Discussion by Sorry-Im-Stupid** [bnpr/Malt#111] [Q&A]
On the subject of noise, in the visual style I'm going for, a certain amount of granularity is necessary, and I've figured out how to achieve the effect with math nodes. But my overall method has several unfortunate limitations. For one thing, only .jpeg exports retain the look I want. The .png exports lose all the good stuff and leave a bunch of anomalies instead. See below:
JPG:
https://user-images.githubusercontent.com/84992987/129072931-b7208ccd-711a-430e-acb9-6295b07f07af.jpg
PNG:
https://user-images.githubusercontent.com/84992987/129072908-a1a123fd-f89f-4ef4-9053-203e90224627.png

You'll also notice that the light angle is the same in both, that is from a matcap I made in eevee for this process. But the gradiant is always relative to the view, or the camera, which again presents some limitations. For one, having to make a custom matcap for every single scene, character, item, and time of day in my projects will eat up more time than setting the actual scenes.
Surely there must be a way to generate the color gradients in malt and have them act as light-relative, or at least make them adjustable in-engine to suit whatever they're applied to.

**Comment by pragma37** [comment]
>Surely there must be a way to generate the color gradients in malt and have them act as light-relative, or at least make them adjustable in-engine to suit whatever they're applied to.

get_diffuse, get_diffuse_half, get_diffuse_gradient, get_specular and get_specular_gradient is what you are looking for.

Regarding the JPG vs PNG issue, my guess is that, since you are setting the alpha on non-transparent objects (https://github.com/bnpr/Malt/discussions/109#discussioncomment-1155929), they look different because by default PNGs are RGBA in Blender but JPGs are RGB.

**Reply by Sorry-Im-Stupid** [reply]
You were right about the alpha channel interference. Interesting, and good to know. But I've tried to get the gradients to work, and it just kinda went black.
![test eevee matcap malt shader4](https://user-images.githubusercontent.com/84992987/129218634-cbd7a7af-be42-416c-bc4c-cf32fb8d0880.png)

**Reply by pragma37** [reply]
You need to add an actual gradient to the node. Please, check the sample file (#94).
And a light, obviously.

**Reply by Sorry-Im-Stupid** [reply]
Aha, the problem was the sun lamp. Improperly set up, as per my fiddling. Is there any chance of a blackbody function for lights some time? I was trying to recreate it, and clearly failing. Major derp on my part.
Thank you, it now works, and looks preeeety close to what I'm going for.
![test eevee matcap malt shader5](https://user-images.githubusercontent.com/84992987/129254834-a5442efc-8471-4cd0-bf50-b10383c1520c.jpg)
